## Image Classification
SVM vs Logistic Regression

In [6]:
import multiprocessing
from sklearn.model_selection import GridSearchCV
from sklearn import svm

In [8]:
cpu_cnt = multiprocessing.cpu_count()
print(cpu_cnt)

4


In [5]:
params = [
    {
    'kernel':['linear','rbf','poly','sigmoid'],
    'C' : [0.1,0.2,0.4,0.5,1.0,2.0,5.0]
    }
]

In [10]:
svc = svm.SVC()
gs = GridSearchCV(estimator=svc,param_grid = params,scoring='accuracy',cv=5,n_jobs=cpu_cnt)

### Load the Dataset

In [12]:
from sklearn.datasets import load_digits

In [13]:
digits = load_digits()

In [17]:
X = digits.data
Y  = digits.target
print(X.shape)
print(Y.shape)

(1797, 64)
(1797,)


## Logistic Regression

In [19]:
from sklearn.model_selection import cross_val_score

In [20]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
cross_val_score(lr,X,Y,scoring="accuracy",cv=5).mean()

0.9210188113360701

Logistic Regresion gives around 92% accuracy!

## Support Vector Machines

In [22]:
svc = svm.SVC()
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()


0.44878680061604637

Pretty Bad Result :(

## Let's try changing HyperParameters using Grid Search

In [23]:
gs.fit(X,Y)### Grid search on SVM Parameters!


GridSearchCV(cv=5, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=4,
       param_grid=[{'kernel': ['linear', 'rbf', 'poly', 'sigmoid'], 'C': [0.1, 0.2, 0.4, 0.5, 1.0, 2.0, 5.0]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=0)

In [24]:
gs.best_estimator_


SVC(C=0.1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [25]:
gs.best_score_

0.9699499165275459

SVM Gives 97% Accuracy using C=0.1, and Polynomial Kernel of Degree 3